File: frontend/src/types/group-trip.ts
export interface GroupTripMember {
  user_id: string;
  name: string;
  picture_url: string | null;
  role: "owner" | "member";
  joined_at: string;
}

export interface GroupTrip {
  id: string;
  creator_id: string;
  title: string;
  destination: string;
  description: string | null;
  start_date: string;
  end_date: string;
  visibility: "public" | "private";
  invite_code: string;
  member_count: number;
  itinerary_id: string | null;
  created_at: string;
  creator_name: string;
  creator_picture_url: string | null;
}

export interface GroupTripDetail extends GroupTrip {
  members: GroupTripMember[];
}

export interface GroupTripPreview {
  id: string;
  title: string;
  destination: string;
  description: string | null;
  start_date: string;
  end_date: string;
  visibility: "public" | "private";
  member_count: number;
  itinerary_id: string | null;
  creator_name: string;
  creator_picture_url: string | null;
}

export interface OverlappingTrip {
  id: string;
  title: string;
  destination: string;
  description: string | null;
  start_date: string;
  end_date: string;
  visibility: "public" | "private";
  member_count: number;
  creator_name: string;
  creator_picture_url: string | null;
}

export interface CreateGroupTripPayload {
  title: string;
  destination: string;
  description?: string;
  start_date: string;
  end_date: string;
  visibility: "public" | "private";
}

export interface UpdateGroupTripPayload {
  title?: string;
  destination?: string;
  description?: string;
  start_date?: string;
  end_date?: string;
  visibility?: "public" | "private";
}

export interface GroupTripListParams {
  page?: number;
  per_page?: number;
}

File: backend/app/schemas/group_trip.py
import uuid
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, model_validator


class GroupTripCreate(BaseModel):
    title: str
    destination: str
    description: Optional[str] = None
    start_date: date
    end_date: date
    visibility: str = "private"  # public | private

    @model_validator(mode="after")
    def validate_dates(self):
        if self.end_date < self.start_date:
            raise ValueError("end_date must be on or after start_date")
        if self.visibility not in ("public", "private"):
            raise ValueError("visibility must be 'public' or 'private'")
        return self


class GroupTripUpdate(BaseModel):
    title: Optional[str] = None
    destination: Optional[str] = None
    description: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    visibility: Optional[str] = None


class MemberRead(BaseModel):
    user_id: uuid.UUID
    name: str
    picture_url: Optional[str] = None
    role: str
    joined_at: datetime

    model_config = ConfigDict(from_attributes=True)


class GroupTripRead(BaseModel):
    id: uuid.UUID
    creator_id: uuid.UUID
    title: str
    destination: str
    description: Optional[str] = None
    start_date: date
    end_date: date
    visibility: str
    invite_code: str
    member_count: int = 0
    itinerary_id: Optional[uuid.UUID] = None
    created_at: datetime

    # Creator info
    creator_name: str
    creator_picture_url: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class GroupTripDetail(GroupTripRead):
    members: List[MemberRead] = []


class GroupTripList(BaseModel):
    items: List[GroupTripRead]
    total: int
    page: int
    per_page: int
    total_pages: int


class GroupTripPreview(BaseModel):
    """Basic info shown to someone deciding whether to join via invite link."""
    id: uuid.UUID
    title: str
    destination: str
    description: Optional[str] = None
    start_date: date
    end_date: date
    visibility: str
    member_count: int = 0
    itinerary_id: Optional[uuid.UUID] = None
    creator_name: str
    creator_picture_url: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class OverlappingTrip(BaseModel):
    id: uuid.UUID
    title: str
    destination: str
    description: Optional[str] = None
    start_date: date
    end_date: date
    visibility: str
    member_count: int = 0
    itinerary_id: Optional[uuid.UUID] = None
    creator_name: str
    creator_picture_url: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)
